swarm.go

```diff
@@ -101,23 +101,40 @@ func NewSwarm(config *api.Config, mockStore *mock.NodeStore) (self *Swarm, err e
 		return nil, fmt.Errorf("empty bzz key")
 	}
 
-	var backend chequebook.Backend
-	if config.SwapAPI != "" && config.SwapEnabled {
-		log.Info("connecting to SWAP API", "url", config.SwapAPI)
-		backend, err = ethclient.Dial(config.SwapAPI)
-		if err != nil {
-			return nil, fmt.Errorf("error connecting to SWAP API %s: %s", config.SwapAPI, err)
-		}
-	}
-
 	self = &Swarm{
 		config:       config,
-		backend:      backend,
 		privateKey:   config.ShiftPrivateKey(),
 		cleanupFuncs: []func() error{},
 	}
 	log.Debug("Setting up Swarm service components")
 
+	// Swap initialization
+	if config.SwapEnabled {
+		// for now, Swap can only be enabled in a whitelisted network
+		if self.config.NetworkID != swap.AllowedNetworkID {
+			return nil, fmt.Errorf("swap can only be enabled under Network ID %d, found Network ID %d instead", swap.AllowedNetworkID, self.config.NetworkID)
+		}
+		// if Swap is enabled, we MUST have a contract API
+		if self.config.BackendURL == "" {
+			return nil, errors.New("swap enabled but no contract address given; fatal error condition, aborting")
+		}
+		log.Info("connecting to SWAP API", "url", self.config.BackendURL)
+		self.backend, err = ethclient.Dial(self.config.BackendURL)
+		if err != nil {
+			return nil, fmt.Errorf("error connecting to SWAP API %s: %s", self.config.BackendURL, err)
+		}
+
+		// initialize the balances store
+		balancesStore, err := state.NewDBStore(filepath.Join(config.Path, "balances.db"))
+		if err != nil {
+			return nil, err
+		}
+		// create the accounting objects
+		self.swap = swap.New(balancesStore, self.privateKey, self.config.Contract, self.backend)
```

Comment:
you should know:
1. in simulations, if you bootstrap `swap.New` through a service function, the simulation will call the service' `Stop()` function.
2. here, you call in `Swarm.Stop()` the `Close()` method. as a result you might be potentially be hitting errors in tests (close of closed channel etc)

Reply:
So please make sure to always call `Stop()` method on shutdown - so correct the shutdown here too

Reply:
See #1664.